File: daemon/src/core/keyed_digest.py
"""Foundation keyed lookup digests and the semantic text transform.

The ``bmas-digest`` profile from the digest utility stays the one
shared content digest. This module adds two separate values:

1. One semantic canonical text value for text comparison. It applies
   NFC and LF line endings and preserves all other whitespace. It
   never replaces the exact content digest.
2. One keyed HMAC-SHA-256 digest for equality checks on sensitive
   values. A plain hash of a sensitive value can still disclose
   information, so equality lookups use one per-tenant key and record
   its key identifier. A keyed digest never appears in a public
   export.

Key rotation keeps the previous epoch queryable, so an equality
lookup checks the active and the previous key epochs.
"""
from __future__ import annotations

import hashlib
import hmac
import unicodedata
from dataclasses import dataclass

from core.digest_profile import (
    DIGEST_PROFILE,
    DIGEST_PROFILE_VERSION,
    DigestInputError,
)

KEYED_DIGEST_ALGORITHM = "hmac-sha256"
KEYED_DIGEST_DOMAIN_PREFIX = "bmas:"


class KeyedDigestError(ValueError):
    """One keyed digest rule failed closed."""


def semantic_text(value: str) -> str:
    """Return the separate canonical text value for semantic comparison.

    The transform applies NFC normalization and LF line endings. It
    preserves every other whitespace character. The result supports
    semantic comparison only; the exact content digest still uses the
    exact validated UTF-8 bytes.
    """
    try:
        value.encode("utf-8")
    except UnicodeEncodeError as exc:
        raise DigestInputError(
            "The semantic text transform rejects invalid Unicode"
        ) from exc
    normalized = unicodedata.normalize("NFC", value)
    return normalized.replace("\r\n", "\n").replace("\r", "\n")


@dataclass(frozen=True)
class TenantKeyEpoch:
    """One tenant HMAC key epoch."""

    tenant_id: str
    key_id: str
    key_bytes: bytes
    epoch: int


@dataclass(frozen=True)
class KeyedDigestRecord:
    """One keyed equality digest with its complete metadata."""

    profile: str
    profile_version: str
    domain: str
    algorithm: str
    value: str
    key_id: str

    def to_dict(self) -> dict[str, str]:
        return {
            "profile": self.profile,
            "profile_version": self.profile_version,
            "domain": self.domain,
            "algorithm": self.algorithm,
            "value": self.value,
            "key_id": self.key_id,
        }


class TenantKeyRing:
    """Per-tenant HMAC keys with active and previous epochs."""

    def __init__(self) -> None:
        self._epochs: dict[str, list[TenantKeyEpoch]] = {}

    def install_key(
        self, tenant_id: str, key_id: str, key_bytes: bytes,
    ) -> TenantKeyEpoch:
        """Install one new active key epoch for a tenant."""
        if len(key_bytes) < 32:
            raise KeyedDigestError(
                "A tenant HMAC key uses at least 32 bytes"
            )
        history = self._epochs.setdefault(tenant_id, [])
        epoch = TenantKeyEpoch(
            tenant_id=tenant_id,
            key_id=key_id,
            key_bytes=key_bytes,
            epoch=len(history) + 1,
        )
        history.append(epoch)
        return epoch

    def active_epoch(self, tenant_id: str) -> TenantKeyEpoch:
        history = self._epochs.get(tenant_id)
        if not history:
            raise KeyedDigestError(
                f"No HMAC key exists for tenant {tenant_id!r}"
            )
        return history[-1]

    def lookup_epochs(self, tenant_id: str) -> list[TenantKeyEpoch]:
        """Return the active and previous epochs for rotation lookup."""
        history = self._epochs.get(tenant_id)
        if not history:
            raise KeyedDigestError(
                f"No HMAC key exists for tenant {tenant_id!r}"
            )
        return list(history[-2:])


def _framed(domain: str, payload: bytes) -> bytes:
    if not domain:
        raise KeyedDigestError("A keyed digest names its domain")
    return (
        KEYED_DIGEST_DOMAIN_PREFIX.encode("ascii")
        + domain.encode("ascii")
        + b"\x00"
        + payload
    )


def keyed_digest(
    ring: TenantKeyRing,
    tenant_id: str,
    domain: str,
    value: str,
) -> KeyedDigestRecord:
    """Digest one sensitive value under the tenant's active key."""
    epoch = ring.active_epoch(tenant_id)
    mac = hmac.new(
        epoch.key_bytes,
        _framed(domain, value.encode("utf-8")),
        hashlib.sha256,
    )
    return KeyedDigestRecord(
        profile=DIGEST_PROFILE,
        profile_version=DIGEST_PROFILE_VERSION,
        domain=domain,
        algorithm=KEYED_DIGEST_ALGORITHM,
        value=mac.hexdigest(),
        key_id=epoch.key_id,
    )


def keyed_lookup_values(
    ring: TenantKeyRing,
    tenant_id: str,
    domain: str,
    value: str,
) -> list[KeyedDigestRecord]:
    """Digest one value under the active and previous key epochs.

    During rotation, an equality lookup queries both epochs, so a
    record written under the previous key still matches.
    """
    records = []
    for epoch in ring.lookup_epochs(tenant_id):
        mac = hmac.new(
            epoch.key_bytes,
            _framed(domain, value.encode("utf-8")),
            hashlib.sha256,
        )
        records.append(
            KeyedDigestRecord(
                profile=DIGEST_PROFILE,
                profile_version=DIGEST_PROFILE_VERSION,
                domain=domain,
                algorithm=KEYED_DIGEST_ALGORITHM,
                value=mac.hexdigest(),
                key_id=epoch.key_id,
            ),
        )
    return records


def export_digest_for_public_view(record: dict[str, str]) -> dict[str, str]:
    """Strip keyed digests from one public export record.

    A keyed digest supports internal equality only. A public export
    keeps unkeyed content digests and drops every keyed value.
    """
    if record.get("algorithm") == KEYED_DIGEST_ALGORITHM or (
        record.get("key_id")
    ):
        return {
            "profile": record.get("profile", DIGEST_PROFILE),
            "domain": record.get("domain", ""),
            "redacted": "keyed_digest_withheld",
        }
    return dict(record)


_REQUIRED_AUDIT_FIELDS = ("profile", "profile_version", "domain",
                          "algorithm", "value")


def audit_digest_record(record: dict[str, str]) -> None:
    """Audit one persisted digest record or fail closed.

    Every shared digest declares its profile, domain, algorithm,
    value, and optional key identifier. A keyed algorithm requires the
    key identifier; an unkeyed algorithm forbids it.
    """
    missing = [name for name in _REQUIRED_AUDIT_FIELDS if not record.get(name)]
    if missing:
        raise KeyedDigestError(f"The digest record misses {missing}")
    if record["profile"] != DIGEST_PROFILE:
        raise KeyedDigestError(
            f"Unknown digest profile: {record['profile']!r}"
        )
    if record["algorithm"] not in ("sha256", KEYED_DIGEST_ALGORITHM):
        raise KeyedDigestError(
            f"Unknown digest algorithm: {record['algorithm']!r}"
        )
    if record["algorithm"] == KEYED_DIGEST_ALGORITHM:
        if not record.get("key_id"):
            raise KeyedDigestError(
                "A keyed digest records its key identifier"
            )
    elif record.get("key_id"):
        raise KeyedDigestError(
            "An unkeyed digest carries no key identifier"
        )

File: mission-control/src/lib/keyed-digest.ts
/**
 * The keyed digest, the semantic text transform, and the exact content
 * digest in TypeScript.
 *
 * The daemon is the reference implementation. Every digest input
 * starts with the frame `bmas:<domain>` and one NUL byte. The keyed
 * digest is an HMAC-SHA256 over that frame under one tenant key. The
 * semantic text transform applies NFC normalization and LF line
 * endings for comparison only. The exact content digest hashes the
 * exact UTF-8 bytes with no normalization.
 *
 * The frozen vectors at daemon/tests/fixtures/keyed_digest.json prove
 * that this implementation and the Python implementation produce
 * byte-identical output.
 */
import { createHash, createHmac } from "node:crypto";

export const KEYED_DIGEST_ALGORITHM = "hmac-sha256";
export const KEYED_DIGEST_DOMAIN_PREFIX = "bmas:";

// A lone surrogate is a high surrogate without a following low surrogate,
// or a low surrogate without a preceding high surrogate.
const HIGH_SURROGATES = "\\ud800-\\udbff";
const LOW_SURROGATES = "\\udc00-\\udfff";
const LONE_SURROGATE = new RegExp(
  `[${HIGH_SURROGATES}](?![${LOW_SURROGATES}])|(?<![${HIGH_SURROGATES}])[${LOW_SURROGATES}]`,
);
const ASCII_DOMAIN = /^[\x21-\x7e]+$/;

export class KeyedDigestError extends Error {}

/** Return the canonical text value for semantic comparison. */
export function semanticText(value: string): string {
  if (LONE_SURROGATE.test(value)) {
    throw new KeyedDigestError("The semantic text transform rejects invalid Unicode");
  }
  return value.normalize("NFC").replace(/\r\n/g, "\n").replace(/\r/g, "\n");
}

/** Frame one payload under one domain: `bmas:<domain>` + NUL + payload. */
export function framedBytes(domain: string, payload: Uint8Array): Uint8Array {
  if (!domain || !ASCII_DOMAIN.test(domain)) {
    throw new KeyedDigestError(`Invalid digest domain: ${domain}`);
  }
  const frame = new TextEncoder().encode(`${KEYED_DIGEST_DOMAIN_PREFIX}${domain}`);
  const framed = new Uint8Array(frame.length + 1 + payload.length);
  framed.set(frame, 0);
  framed[frame.length] = 0;
  framed.set(payload, frame.length + 1);
  return framed;
}

/** SHA-256 of the exact framed bytes, as lowercase hex. */
export function exactBytesDigestHex(domain: string, payload: Uint8Array): string {
  return createHash("sha256").update(framedBytes(domain, payload)).digest("hex");
}

/** SHA-256 of the exact UTF-8 bytes of one text value, framed under one domain. */
export function exactTextDigestHex(domain: string, value: string): string {
  if (LONE_SURROGATE.test(value)) {
    throw new KeyedDigestError("The exact content digest rejects invalid Unicode");
  }
  return exactBytesDigestHex(domain, new TextEncoder().encode(value));
}

/** HMAC-SHA256 of one framed text value under one tenant key, as lowercase hex. */
export function keyedDigestHex(keyBytes: Uint8Array, domain: string, value: string): string {
  if (keyBytes.length < 32) {
    throw new KeyedDigestError("A keyed digest key needs at least 32 bytes");
  }
  if (LONE_SURROGATE.test(value)) {
    throw new KeyedDigestError("The keyed digest rejects invalid Unicode");
  }
  const framed = framedBytes(domain, new TextEncoder().encode(value));
  return createHmac("sha256", keyBytes).update(framed).digest("hex");
}

export function hexToBytes(hex: string): Uint8Array {
  if (hex.length % 2 !== 0 || !/^[0-9a-f]*$/.test(hex)) {
    throw new KeyedDigestError("A key is lowercase hex with an even length");
  }
  const bytes = new Uint8Array(hex.length / 2);
  for (let index = 0; index < bytes.length; index += 1) {
    bytes[index] = Number.parseInt(hex.slice(index * 2, index * 2 + 2), 16);
  }
  return bytes;
}
